[DataPipe] Simple graph snapshotting

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79479

This mostly completes the "poor man's snapshotting" implementation (named "simple snapshotting"). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` of the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

Differential Revision: [D37943406](https://our.internmc.facebook.com/intern/diff/D37943406)

NivekT committed:
[DataPipe] Fast-forwarding feature

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fast-forwarding feature"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fast-forwarding feature"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Fast-forwarding feature"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph fast-forwarding"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph fast-forwarding"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph fast-forwarding"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph fast-forwarding"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph fast-forwarding"




[ghstack-poisoned]

ejguan reviewed (commented):
This is the question I asked before. Do we need to make DataPipe itself supporting fast forward.
If we directly use `DataLoader2` to save/restore the state, we can simply fast forward the iterator of DataPipe graph directly in DataLoader2.

NivekT commented:
> This is the question I asked before. Do we need to make DataPipe itself supporting fast forward. If we directly use `DataLoader2` to save/restore the state, we can simply fast forward the iterator of DataPipe graph directly in DataLoader2.

I discuss part of the design [here](https://github.com/pytorch/pytorch/pull/80250).
1. For the most basic implementation like the one here, `DataLoader2` can directly call the fast forward function in this PR and and get the iterator to the desired state. However, this requires re-doing all the computations.
2. For more efficient implementations (in a future PR), `DataLoader2` will still call a fast forward API, but individual DataPipes will need custom fast-forward methods that allow efficient restoration state.
    - Notably, 1) stateless DataPipes do not need to do anything, 2) stateful ones may be able to restore from a serialized buffer, 3) source DataPipes (e.g. `IterableWrapper`) may still do a simple fast forward

ejguan commented:
> They serve the same purpose. I can rename fast-forward as "poor man's snapshotting" if that is better.

Could we combine these two APIs as the single one?

NivekT commented:
> Could we combine these two APIs as the single one?

My vision is that `DataLoader2` will have a single API to call the relevant functions to save/restore snapshot. The specific methods used for each DataPipe will need to be determined on the fly (depends on what DataPipes the graph has). There will not be duplicate, I can make sure of that. I will also make these functions private for now until there is a good reason for users to call them directly.

VitalyFedyunin commented on torch/utils/data/datapipes/utils/snapshot.py:
```diff
@@ -0,0 +1,47 @@
+            should be in its `initial` state as it was first passed into ``DataLoader`` or ``ReadingService``.
+    """
+    apply_shuffle_seed(datapipe, rng)
+
```
You should probably check if DP is already in iterator stage and prevent fast forwarding it. This will make sure that snapshot can be only restored to the 'new' graph.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2703,5 +2706,59 @@ def reset(self):
+        initial_rng_state = rng.get_state()
+        torch.utils.data.graph_settings.apply_shuffle_seed(datapipe, rng)
+        expected_full_res = list(datapipe)
+
```
Change rng state here (to something totally different) and do one more 'epoch' to make sure we can restore state between epochs.

NivekT replied:
I just added that. Please have a look to see if it is what you meant

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2703,5 +2706,59 @@ def reset(self):
+        graph7 = cdp1.concat(cdp2)
+        self._fast_forward_graph_test_helper(graph7, simple_fast_forward_graph, rng=rng)
+
+
```
Missing tests:

What will happen if I get state, restore it on the new graph, do extra 5 iterations, save state and restore it on the new graph.

NivekT replied:
This is in the next PR.

NivekT replied:
I combined that PR with this one.

NivekT committed:
Update on "[DataPipe] Simple graph fast-forwarding"




[ghstack-poisoned]

NivekT commented:
Note that I squashed two commits together and updated this PR.

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple fast forward). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple fast-forward (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_fastforward = deserialized_graph._number_of_samples_yielded
simple_fast_forward_graph(deserialized_graph, n_fastforward, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

If this looks acceptable, I can modify `DataLoader2` to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.

In the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple fast-forward is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple fast forward). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple fast-forward (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_fastforward = deserialized_graph._number_of_samples_yielded
simple_fast_forward_graph(deserialized_graph, n_fastforward, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

If this looks acceptable, I can modify `DataLoader2` to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.

In the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple fast-forward is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple fast forward). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple fast-forward (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_fastforward = deserialized_graph._number_of_samples_yielded
simple_fast_forward_graph(deserialized_graph, n_fastforward, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

If this looks acceptable, I can modify `DataLoader2` to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.

In the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple fast-forward is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT commented on test/test_datapipe.py:
```diff
@@ -2755,5 +2758,157 @@ def reset(self):
         self.assertEqual(3, datapipe._number_of_samples_yielded)
 
+
+class _CustomNonGeneratorTestDataPipe(IterDataPipe):
```
Creating these classes outside of the test so that they can be serialized

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -371,9 +371,12 @@ def wrap_setstate(*args, **kwargs):
-                datapipe = args[0]
-                datapipe._restored = True
-                return setstate_func(*args, **kwargs)
+                try:
+                    return setstate_func(*args, **kwargs)
+                finally:
+                    # This needs happen after because the state being passed in may have `datapipe._restored = False`.
+                    datapipe = args[0]
+                    datapipe._restored = True
```
Rewriting the `__setstate__` hook logic because the previous version doesn't work with serialization. In particular, `datapipe._restored` is set to `False` because `setstate_func` overwrites it.

NivekT commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -154,6 +155,12 @@ def __getstate__(self):
             return IterDataPipe.getstate_hook(self)
         return self.__dict__
 
+    def __setstate__(self, state):
+        """
+        This method definition is necessary so that the hook function in `_typing.py` can work.
+        """
+        self.__dict__ = state
```
I realized this is needed for the hook to work. Let me know if there is a better idea.

NivekT commented on torch/utils/data/datapipes/utils/snapshot.py:
```diff
@@ -0,0 +1,48 @@
+#   2. `in_batch_shuffle` and `bucketbatch` are not compatible with this because they currently
+#      lack the option to `set_seed`.
+
+def _simple_snapshot_graph(datapipe: IterDataPipe, n_iterations: int, rng=None) -> None:
```
The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that iterator to the DataPipe
4. The next time `iter` is called on the DataPipe, use the forwarded iterator

ejguan replied:
What would be the expected behavior if we:
1. Create a iterator
2. load snapshot
3. continue iterate over the iterator

NivekT replied:
Creating an iterator will set `datapipe._restored = True`, and you can't load snapshot after, because of the condition here:
https://github.com/pytorch/pytorch/blob/7a74a5c1d91c4c4c968e0953dc5a0cc7520be48f/torch/utils/data/datapipes/utils/snapshot.py#L27-L30

Let me add a warning there (let me know if you prefer it to be an exception).

ejguan replied:
If that's the case, I think raising an Error is a more proper solution.

ejguan commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -161,7 +168,7 @@ def wrap_generator(*args, **kwargs):
                         response = gen.send(request)
             except StopIteration as e:
-                return e.value
+                return e.value  # noqa: B901
```
Do we need to return anything from here?

NivekT replied:
actually nope

ejguan commented on torch/utils/data/datapipes/utils/snapshot.py:
```diff
@@ -0,0 +1,48 @@
+    """
+    # This operation can only take place if `datapipe` has been recently restored (usually through `__setstate__`)
+    # and an iterator has not started yet.
+    if not datapipe._restored:
+        return
```
I am still confused here. `_restored` indicates the DataPipe is restored from snapshot.
Then, after the DataPipe is restored, we still want to do fast forwarding here. Why not we directly fast-forward when restoring from snapshot?

NivekT replied:
Perhaps we should rename `_restored`, currently two built-in things can set `_restored = True`:
1. `__setstate__` (usually called by deserialization)
2. The function that we are adding in this PR

NivekT replied:
I think perhaps we need 3 states:
1. `Deserialized` - allows you to restore snapshot and create an iterator without reset
2. `Restored` - cannot restore again, allows you to create an iterator without reset
3. `Iterating`? - cannot restore, will reset if you create a new iterator

ejguan replied:
For the state 3, we should allow restore during iterating. I think normally we would do:
```py
it = iter(dl)
for _ in range(10):
    next(it)

state = dl.state_dict()

dl.restore(state  # Restore during iteration?
it = iter(dl)
```

NivekT replied:
I think it is feasible to support this, but how common is this and why is this done? 
Note that the original `iterator` will be invalidated during the restoration.

@VitalyFedyunin Any thoughts?

NivekT replied:
Note that I updated the PR with the 3 states as discussed above. Thanks for the feedback!

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple snapshot). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` pf the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple snapshot). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` pf the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

ejguan commented on test/test_datapipe.py:
```diff
@@ -2755,5 +2759,161 @@ def reset(self):
+
+class TestIterDataPipeGraphFastForward(TestCase):
+
+    def _fast_forward_graph_test_helper(self, datapipe, fast_forward_fn, n_iterations=3, rng=None):
```
Do we need `fast_forward_fn` as argument? It seems all of DataPipes should use the same one.

NivekT replied:
For now, no. I was planning to add other function in the upcoming PRs.

ejguan commented on test/test_datapipe.py:
```diff
@@ -2755,5 +2759,161 @@ def reset(self):
+        graph7 = cdp1.concat(cdp2)
+        self._fast_forward_graph_test_helper(graph7, _simple_graph_snapshot_restoration, rng=rng)
+
+        # Raises an exception if the graph has already been restored
+        with self.assertRaisesRegex(RuntimeError, "Snapshot restoration cannot be applied."):
+            _simple_graph_snapshot_restoration(graph7, 1)
```
IIRC, what we have decided is we don't allow restoring from snapshot during iteration. It seems we are not even allowing to restore from snapshot multiple times even if the iteration has not started. Is this intentional?

NivekT replied:
That's what I have in mind, because it doesn't to me why you would want to restore the same DataPipe (or graph) multiple times in a row? It seems like a potential user error?
But this is up for discussion and I'm open to change this and allow it to happen

ejguan replied:
I am basically trying to align the behavior between DataPipe and DataLoader2: https://github.com/pytorch/data/blob/f629899aece23a15c155c7c7805b8db7a2bed2a7/torchdata/dataloader2/dataloader2.py#L176-L179

NivekT replied:
So basically in DL2 you can load as many times as you want as long as you never created an iterator?
I guess my question is what is the use case for loading multiple times and not using it?

NivekT replied:
Perhaps the middle ground is to allow it, but raises a warning instead? Because I think if someone is restoring multiple snapshots without iterating, they may be making a mistake.

NivekT replied:
Since we are going to change DL2 to prevent this behavior, I am not going to make a change here (leaving the exception as it is). Let me know if I should still put a warning instead.

NivekT commented on test/test_datapipe.py:
```diff
@@ -2755,5 +2759,161 @@ def reset(self):
+class TestIterDataPipeGraphFastForward(TestCase):
+
+    def _fast_forward_graph_test_helper(self, datapipe, fast_forward_fn, n_iterations=3, rng=None):
+        datapipe._snapshot_state = _SnapshotState.Deserialized
```
In this test case, there is no serialization/deserialization. The next test does.

NivekT commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -154,6 +156,12 @@ def __getstate__(self):
+        """
+        This method definition is necessary so that the hook function in `_typing.py` can work.
+        """
+        self.__dict__ = state
```
I think not all `super()` has `__setstate__`. I can change it to your first suggestion.

ejguan replied:
You can do a quick experiment. I believe Python class always has a `__setstate__` function hidden from the scene

NivekT replied:
It doesn't seem to be the case. I changed the function to:
```python
def __setstate__(self, state):
    """
    This method definition is necessary so that the hook function in `_typing.py` can work.
    """
    return super().__setstate__(state)
```
Then, I ran this code snippet:
```python
dp = IterableWrapper(range(10))
serialized_dp = pickle.dumps(dp)
deserialized_dp = pickle.loads(serialized_dp)
print(list(dp))
```
It raises:
```
AttributeError: 'super' object has no attribute '__setstate__'
```

NivekT replied:
I think `self.__dict__.update(state)` is likely the best option (aside from `self.__dict__ = state`)

NivekT commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -111,7 +112,8 @@ class IterDataPipe(IterableDataset[T_co], metaclass=_IterDataPipeMeta):
     _valid_iterator_id: Optional[int] = None
     _number_of_samples_yielded: int = 0
-    _restored: bool = False
+    _snapshot_state: _SnapshotState = _SnapshotState.Iterating
```
Actually we can change `Deserialized` to `NotStarted` and make that the default as well.

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple snapshot). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` pf the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the poor man's snapshotting implementation (named simple snapshot). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` pf the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -206,6 +227,11 @@ def wrap_next(*args, **kwargs):
+            if datapipe._fast_forward_iterator:
+                iter_ret = datapipe._fast_forward_iterator
+                datapipe._fast_forward_iterator = None
+                datapipe._snapshot_state = _SnapshotState.Iterating
```
I believe it is. When `_set_datapipe_valid_iterator_id` in the next line, it should trigger `reset`, which should set `datapipe._snapshot_state = _SnapshotState.Iterating`.

NivekT commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -206,6 +227,11 @@ def wrap_next(*args, **kwargs):
+                iter_ret = datapipe._fast_forward_iterator
+                datapipe._fast_forward_iterator = None
+                datapipe._snapshot_state = _SnapshotState.Iterating
+                return iter_ret
```
Discussed offline. Changing the implementation to allow restoration from `Iterating` state.

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the "poor man's snapshotting" implementation (named "simple snapshotting"). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` of the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37943406).

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the "poor man's snapshotting" implementation (named "simple snapshotting"). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` of the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

Differential Revision: [D37943406](https://our.internmc.facebook.com/intern/diff/D37943406)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37943406).

NivekT commented on test/test_datapipe.py:
```diff
@@ -2983,5 +2987,160 @@ def reset(self):
+
+class TestIterDataPipeGraphFastForward(TestCase):
+
+    def _fast_forward_graph_test_helper(self, datapipe, fast_forward_fn, n_iterations=3, rng=None):
```
Done

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -368,12 +371,22 @@ def conditional_reset(*args, **kwargs):
-                datapipe = args[0]
-                datapipe._restored = True
-                return setstate_func(*args, **kwargs)
+                try:
+                    return setstate_func(*args, **kwargs)
+                finally:
+                    # This needs happen after because the state being passed in
+                    # may have `datapipe._SnapshotState = Iterating`.
+                    datapipe = args[0]
+                    # FIXME: Without the `if` condition below, the code snippet below would fail. We have
+                    #   investigated but no clue why. Note that this doesn't cause any functional issue.
+                    #   >>> dp = IterableWrapper(range(10))
+                    #   >>> copied_dp = pickle.loads(pickle.dumps(dp))
+                    #   >>> pickle.dumps(dp) == pickle.dumps(copied_dp)
+                    if datapipe._snapshot_state != _SnapshotState.NotStarted:
+                        datapipe._snapshot_state = _SnapshotState.NotStarted
```
@VitalyFedyunin you may want to take a closer look at this `setstate` hook. Somehow the bytes from `pickle.dumps` changes when we set `datapipe._snapshot_state = _SnapshotState.NotStarted`, even was `datapipe._snapshot_state` was already `_SnapshotState.NotStarted` prior.

This doesn't cause any functional difference/issue.

@ejguan and I tried to look into why but no luck. Please have a quick look if you can.

VitalyFedyunin replied:
That's weird. Have you tried to type check datapipe._snapshot_state after restoration. It might become `int` and be serialized differently next time (just guessing).

NivekT replied:
On a second thought, given we decided that restoration is possible even when `datapipe._snapshot_state == _SnapshotState.Iterating`. This `setstate` hook is no longer necessary. All the tests will still pass as it is even without the hook.

The proposed changed here is - do NOT set `datapipe._snapshot_state = _SnapshotState.NotStarted` during `__setstate__`. When a DataPipe graph gets serialized/deserialized, its `_snapshot_state` will not change.

cc: @ejguan

NivekT reviewed (commented):
Updated this PR. I am expecting this to be final.

NivekT commented on test/test_datapipe.py:
```diff
@@ -2987,5 +2990,180 @@ def reset(self):
+        """
+        if rng is None:
+            rng = torch.Generator()
+        rng.manual_seed(0)
```
Originally it was randomly generated but @ejguan commented that it would be better to hard-code the expected results. I can't use a random seed if the expected result is hard-coded. Perhaps we can test both.

cc: @ejguan

VitalyFedyunin replied:
Ok, makes sense (and we can leave it as as). But, what if we change internal implementation of `shuffle`, leaving it deterministic etc. but the order will change even for seed 0.

NivekT replied:
Okay, I made a small change to make the result slightly less hard-coded for shuffling. It should be able to handle any change to `shuffle`.

NivekT commented on test/test_datapipe.py:
```diff
@@ -2987,5 +2990,180 @@ def reset(self):
+        rng_for_deserialized = torch.Generator()
+        rng_for_deserialized.manual_seed(0)
+        _simple_graph_snapshot_restoration(deserialized_graph, n_iter, rng=rng_for_deserialized)
+        self.assertEqual(expected_res[n_iter:], list(it))
```
Added

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the "poor man's snapshotting" implementation (named "simple snapshotting"). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` of the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

Differential Revision: [D37943406](https://our.internmc.facebook.com/intern/diff/D37943406)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37943406).

ejguan reviewed (approved):
LGTM based on the test cases

NivekT commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -362,6 +368,7 @@ def _find_next(self, instance_id: int) -> T_co:
     def get_next_element_by_instance(self, instance_id: int):
         if self._datapipe_iterator is None and not self.main_datapipe_exhausted:
             self._datapipe_iterator = iter(self.main_datapipe)
+            self._snapshot_state = _SnapshotState.Iterating  # This is necessary for the DataPipe to reset properly.
```
Yes, because creating a new iterator is always allowed but using an invalidated iterator isn not allowed. I believe `self._snapshot_state` can tell us whether or not an iterator is created, but not which iterator is valid.

ejguan replied:
Oh yea. Forget datapipe doesn't keep track of its iterators. NVM

NivekT commented:
@pytorchbot rebase

NivekT committed:
Update on "[DataPipe] Simple graph snapshotting"


This mostly completes the "poor man's snapshotting" implementation (named "simple snapshotting"). This is the most basic version of snapshotting but it should work for all DataPipes. I will be adding more efficient implementation for different types of DataPipes in future PRs.

### Implementation

The general idea of the simple snapshot is that we will:
1. Create a new iterator
2. Move that iterator forward by `n_iterations`
3. Save that as the `_fast_forward_iterator` of the DataPipe
4. The next time `iter` is called on the DataPipe, use the `_fast_forward_iterator`

### Usage
As of this implementation, the usage will something like:
```python
rng = torch.Generator()
initial_rng_state = rng.get_state()
datapipe: IterDataPipe = ...
# Some usage of the DataPipe, here maybe yielding the first 5 values
n_iter = 5
it = iter(datapipe)
for _ in range(n_iter):
    next(it)
serialized_graph = pickle.dumps(datapipe)

# The serialized object has most of the sufficient information for simple snapshot (except for initial RNG state)
# It can be deserialized at a later point in time or by a different process
deserialized_graph = pickle.loads(serialized_graph)
# I think `DataLoader2` or `ReadingService` should store `initial_rng_state` that can be saved by the API that we later use
rng_for_deserialized = torch.Generator()
rng_for_deserialized.set_state(initial_rng_state)
n_iterations = deserialized_graph._number_of_samples_yielded

_simple_snapshot_graph(deserialized_graph, n_iterations, rng=rng_for_deserialized)
# The while DataPipe graph should have the same state as before serialization, such that:
self.assertEqual(list(it), list(deserialized_graph))  # True
```

### Next Steps
If this looks acceptable, the next step is I will modify `DataLoader2`'s prototype ReadingService (the one with queues) to remember things like `initial_rng_state` and to have methods `save_snapshot` that will return the `(serialized graph, initial_rng)` and `restore_snapshot`. This should work for single worker data loading.


Note that, in the long term, `initial_rng_state` may not be necessary if we are able to directly save/restore the buffer and RNG state of `Shuffler` (that is work in progress). However, `initial_rng_state` and simple snapshot is still a good fall-back option for some edge cases where the buffer can't be stored.

Differential Revision: [D37943406](https://our.internmc.facebook.com/intern/diff/D37943406)

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37943406).

NivekT commented:
@pytorchbot merge
